OneSignal/OneSignal-Android-SDK
# [Improve] Handle HTTP header Retry-After from responses from OneSignal

# Description
## One Line Summary
Respect [standard HTTP Retry-After header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) from any responses from OneSignal's backend.

## Details

### Motivation
When OneSignal's backend is under heavy load they want to be able to throttle back requests to them.

### Scope
Only effects delaying follow up network requests after we receive `Retry-After` in a response. If `OpeartionRepo` triggers this then it handles this smartly by delaying future requests, so more batching can be done. Other code that depends on the `HttpClient` simply delays requests to respect this header.

# Testing
## Unit testing
Added and updated a number of Unit tests. OperationRepo, it's executors, and HttpClientTests. 

## Manual testing
Tested on Android 14 emulator. Ensured we correctly handled 429 with it's `Retry-After` by doing the following.
1. In a browser on my machine refreshed /users fetch over and over until I got a `429`
   * https://api.onesignal.com/apps/{{APP_ID_HERE}}/users/by/onesignal_id/{{ONESIGNAL_ID_HERE}}
2. Opened the app on my emulator (using the same internet connection)
3. Check the logcat to ensure the delay was respected.
```
2024-04-24 22:14:24.572  4583-4616  OneSignal               com.onesignal.sdktest                D  [DefaultDispatcher-worker-1] HttpClient: GET apps/77e32082-ea27-42e3-a898-c72e141824ef/users/by/onesignal_id/102c7c89-54f8-49fb-96cb-f55420f1b789
2024-04-24 22:14:24.603  4583-4616  OneSignal               com.onesignal.sdktest                D  [DefaultDispatcher-worker-1] HttpClient: Response Retry-After: 44
2024-04-24 22:14:24.605  4583-4616  OneSignal               com.onesignal.sdktest                D  [DefaultDispatcher-worker-1] HttpClient: GET apps/77e32082-ea27-42e3-a898-c72e141824ef/users/by/onesignal_id/102c7c89-54f8-49fb-96cb-f55420f1b789 - FAILED STATUS: 429
2024-04-24 22:14:24.612  4583-4616  OneSignal               com.onesignal.sdktest                W  [DefaultDispatcher-worker-1] HttpClient: null RECEIVED JSON: {"errors":["API rate limit exceeded"],"limit":"API Per App"}



2024-04-24 22:15:08.839  4583-4618  OneSignal               com.onesignal.sdktest                D  [DefaultDispatcher-worker-3] HttpClient: GET apps/77e32082-ea27-42e3-a898-c72e141824ef/users/by/onesignal_id/102c7c89-54f8-49fb-96cb-f55420f1b789
2024-04-24 22:15:09.010  4583-4618  OneSignal               com.onesignal.sdktest                D  [DefaultDispatcher-worker-3] HttpClient: GET apps/77e32082-ea27-42e3-a898-c72e141824ef/users/by/onesignal_id/102c7c89-54f8-49fb-96cb-f55420f1b789 - STATUS: 200 JSON: 
```

# Affected code checklist

# Checklist
## Overview
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR

## Testing

## Final pass
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/OneSignal/OneSignal-Android-SDK/2064)
